File: Python/HeapSort.py
def swap(arr,i,j):
    temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp

def heapify(arr,i,heapsize):
    max = 0
    l = i*2+1
    r = i*2+2

    if((l<heapsize)and(arr[l]>arr[i])):
        max = l
    else:
        max = i

    if((r<heapsize)and(arr[r]>arr[max])):
        max = r

    swap(arr,i,max)

    if(i!=max):
        heapify(arr,max,heapsize)


def sort(arr):
    heapsize = len(arr)

    top = (heapsize//2)-1

    for i in range(top,-1,-1):
        heapify(arr,i,heapsize)

    for i in range(heapsize-1,0,-1):
        swap(arr,0,i)
        heapsize -= 1
        heapify(arr,0,heapsize)


arr = [123,12,45,1,2,78978,123,-1,-12312,234,88888]
sort(arr)

print(arr)

File: Javascript/HeapSort.js
/* This program implements the heapsort algorithm. heapsort is an in-place sort and its not stable.
   It has worst case time complexity of O(NlgN) */


/*This function constructs a max heap rooted at node index, it starts with the assumption that trees rooted at
   the left-child and the right-child of index are max heaps */
function maxHeapify(arr,heapLength,index)
{
    var maximum = index;
    var leftChild = index*2+1;
    var rightChild = index*2+2;

    if((leftChild<heapLength)&&(arr[leftChild]>arr[index]))
    {
        maximum = leftChild;
    }

    if((rightChild<heapLength)&&(arr[rightChild]>arr[maximum]))
    {
        maximum = rightChild;
    }

    if(maximum!=index)
    {
        var temp = arr[maximum];
        arr[maximum] = arr[index];
        arr[index] = temp;
        maxHeapify(arr,heapLength,maximum);
    }
}

function buildMaxHeap(arr)
{
    var length = arr.length;
    var heapLength = arr.length;
    var startIndex = Math.floor((length-2)/2);

    while(startIndex>=0)
    {
        maxHeapify(arr,heapLength,startIndex--);
    }
}

function heapSort(arr)
{
    buildMaxHeap(arr);

    var startIndex = arr.length-1;
    var heapLength = arr.length;
    while(startIndex>0)
    {
        var temp = arr[startIndex];
        arr[startIndex] = arr[0];
        arr[0] = temp;
        maxHeapify(arr,--heapLength,0);
        startIndex--;
    }
}


var arr = [];

for(var i=100000,j=0;i>0;i--)
{
    arr[j++]=i;

}

heapSort(arr);
console.log("Sorted array : "+arr);